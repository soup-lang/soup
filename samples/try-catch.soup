func asyncThrowFunc(t, i, %%)
    if t
        throw 'this is an asynchronous exception: ' + i

func syncThrowFunc(t, i)
    if t
        throw 'this is a synchronous exception: ' + i

func f()
    try
        asyncThrowFunc(true, 9, %%)
    catch e
        console.log(e)

f()

func g()
    try
        asyncThrowFunc(false, -1, %%)
        [0, 9] |: asyncThrowFunc(false, -1, %%)
        syncThrowFunc(true, 10)
    catch e
        console.log(e)

g()

try
    x: []
    x.push(false)
    x.push(true)
    x |: asyncThrowFunc($, $i, %%)
catch e
    console.log(e)

try
    x: []
    x.push(false)
    x.push(false)
    x.push(true)
    x |: syncThrowFunc($, $i)
catch e
    console.log(e)

try
    try
        x: []
        x.push(true)
        x |: syncThrowFunc($, 0)
    catch e
        setTimeout(%, 0)
        throw 'throw in catch'
catch e
    setTimeout(%, 0)
    console.log(e)

func throwAfterTryCatchSync(t, i)
    try
        console.log('try+')
    catch e
        console.log('catch+')
    syncThrowFunc(t, i)

func throwAfterTryCatchAsync(t, i, %%)
    try
        setTimeout(%, 0)
        console.log('try*')
    catch e
        console.log('catch*')
    syncThrowFunc(t, i)

try
    throwAfterTryCatchSync(true, 72)
catch e
    console.log(e)

try
    throwAfterTryCatchAsync(true, 77, %%)
catch e
    console.log(e)

console.log('end')
