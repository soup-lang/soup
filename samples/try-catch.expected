this is an asynchronous exception: 9
this is a synchronous exception: 10
this is an asynchronous exception: 1
this is a synchronous exception: 2
throw in catch
try+
this is a synchronous exception: 72
try*
this is a synchronous exception: 77
end
